doc/rst/technotes/errorHandling.rst

```diff
@@ -4,190 +4,581 @@
+
+Chapel currently supports three error handling modes: fatal, relaxed, and
+strict. Developers can select each mode in the following ways:
+
+ * *fatal mode*
+
+   * writing code in an implicit module
+   * writing code in a ``prototype module``
+   * passing the ``--permit-unhandled-module-errors`` flag on the
+     ``chpl`` command line
+
+ * *relaxed mode*
+
+   * writing code inside of a ``module`` declaration
+
+ * *strict mode*
+
+   * writing code inside of a module marked with a pragma
+
+It is fully permissible to write for a stricter error handling mode
+while using a looser one -- for example, code that compiles in strict
+mode will also compile in relaxed or fatal mode.
+
+.. _technote-errorHandling-fatal:
+
+Fatal Mode for Implicit and Prototype Modules
++++++++++++++++++++++++++++++++++++++++++++++
+
+In implicit and prototype modules, it is not necessary to explicitly handle
+errors from a function that throws. If an error is thrown and the calling
+function throws, the error will be propagated out of the function.  However,
+if an error is thrown and the calling function does not throw, the program
+will halt.
+
+This is the *fatal* error handling mode.
+
+An *implicit module* is the module the compiler creates to store
+code not in a ``module`` declaration. In particular, all of the
+statements above this point in this file are in an implicit module called
+errorHandling. Since the below statements are also not in a ``module``
+declaration, they are also in an *implicit module*.
+
+.. code-block:: chapel
+
+  canThrow(1); // handling can be omitted; halts if an error occurs
+
+  proc throwsErrorsOn() throws {
+    // error propagates out of this function
+    canThrow(-1);
+  }
+
+  proc doesNotThrowErrorsOn() {
+    // causes a halt if called
+    alwaysThrows();
+  }
+
+Fatal error mode can also be activated for explicit modules using the
+``prototype module`` declaration:
+
+.. code-block:: chapel
+
+  prototype module PrototypeModule {
+
+    canThrow(1); // handling can be omitted; halts if an error occurs
+
+    proc throwsErrorsOn() throws {
+      // error propagates out of this function
+      alwaysThrows();
+    }
+
+    proc doesNotThrowErrorsOn() {
+      // causes a halt if called
+      alwaysThrows();
+    }
+  }
+
+  use PrototypeModule;
+
+.. _technote-errorHandling-relaxed:
+
+Relaxed Mode for Explicit Modules
++++++++++++++++++++++++++++++++++
+
+In non-prototype explicit modules, it is necessary to handle errors if the
+calling function does not throw. If the calling function does
+throw, then the error will be propagated out as with implicit modules.
+
+This is the *relaxed* error handling mode.
+
+.. code-block:: chapel
+
+  module ProductionModule {
+    // this line would cause compilation error since the error is not handled
+    // canThrow(1);
+
+    proc throwsErrorsOn() throws {
+      // any error thrown in thrownErrors will propagates out of this function
+      alwaysThrows();
+    }
```

Comment:
Should we also have a commented out doesNotThrowErrorsOn here, with the comment that it wouldn't compile because the error is not handled?